### ISSUE TITLE
Predicate on column in join equivalence is often not pushed to all inputs

### ISSUE THREAD

**Issue by wangandi** [MaterializeInc/materialize#4411]
`CREATE TABLE foo(a int, b int)`
`CREATE TABLE bar(a int, b int)`

No explicitly created indexes. Due to lack of primary keys specified, the indexes are implicitly `foo(a,b)` and `bar(a,b)`.

```
EXPLAIN PLAN FOR select * from foo inner join bar on foo.a = bar.a where foo.a = 1
----
%0 =
| Get materialize.public.foo (u1)
| Filter !(isnull(#0)), (#0 = 1)
| ArrangeBy (#0)  

%1 =
| Get materialize.public.bar (u3)
| Filter !(isnull(#0)), (#0 = 1)

%2 =
| Join %0 %1 (= #0 #2) 
| | implementation = Differential %1 %0.(#0) 
| | demand = (#0, #1, #3) 
| Project (#0, #1, #0, #3)
```

```
EXPLAIN PLAN FOR select * from foo, bar where foo.a = bar.a and foo.a = 1
%0 =
| Get materialize.public.foo (u1)
| Filter (#0 = 1)
| ArrangeBy (#0)

%1 =
| Get materialize.public.bar (u3)

%2 =
| Join %0 %1 (= #0 #2)
| | implementation = Differential %1 %0.(#0)
| | demand = (#0, #1, #3)
| Project (#0, #1, #0, #3)
```

The filter does not make it into %1 in the latter case, which means that in the join, Materialize will maintain an arrangement on keyed on `a` for the entirety of bar instead of the filtered version of bar.

The reason for the difference in plans is order of operation. 
The order when `inner join` is explicitly stated:
1) foo.a = bar.a becomes an equivalence class in the join
2) A function pushes `foo.a = 1` down into `foo`. The function sees that the join has the equivalence class `foo.a = bar.a`, so `bar.a = 1` is also true.

The order when `inner join` is not explicitly stated:
1) A function pushes `foo.a = 1` down into `foo`. The function does not see any equivalence class in the join.
2) foo.a = bar.a becomes an equivalence class in the join

That said, the plan for the version of the query that uses the explicit `inner join` syntax is suboptimal because of the unnecessary `Filter !(isnull(#0))`, but this should have less of a negative impact compared to the issue described above, and I'll create a separate issue if I can identify the cause.

**Comment by wangandi** [comment]
I tried 
```
EXPLAIN PLAN FOR select * from foo inner join bar on foo.a = abs(bar.a) where mod(foo.a, 2) = 1
```
```
EXPLAIN PLAN FOR select * from foo, bar where foo.a = abs(bar.a) and mod(foo.a, 2) = 1
```
and in both cases, the result is
```
%0 =
| Get materialize.public.foo (u1)
| Filter !(isnull(#0)), (1 = (#0 % 2))
| ArrangeBy (#0) 

%1 =
| Get materialize.public.bar (u3)
| Filter !(isnull(abs(#0)))

%2 =
| Join %0 %1 (= #0 abs(#2))
| | implementation = Differential %1 %0.(#0) 
| | demand = (#0..#3)
```

By getting a `println` into the internals, I see that the order of operations problem is present when the join key is an expression as opposed to a plain column even when one has explicitly stated `inner join` in the query.

**Comment by wangandi** [comment]
Fixed in #5873.
